### PR TITLE
updates for timesync and test

### PR DESCRIPTION
**Pull request by EricChen-Lei**
implementation of time sync and unit testing

# PR Details
## Description

The implementation of  InfrastructureTimeInterface class, which includes encoding time message and updating to registered instances.
Unit testing for InfrastructureTimeInterface

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        infrastructureTimeMessage.set_seq(2);
+        infrastructureTimeMessage.set_timestep((long)20230407);
+
+        infrastructureTimeInterface.onTimestepUpdate(infrastructureTimeMessage);
+
+        Gson gson = new Gson();
+        String temp = gson.toJson(infrastructureTimeMessage);
+        byte[] t_byte = temp.getBytes();
+
+        verify(infrastructureInstanceMock1, times(1)).sendTimesyncMsgs(eq((byte[])t_byte));
```
Comment (paulbourelly999) [inline co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java:59]: You call send message twice, why is this only called once.

Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java:59]: Hi Paul, in this line, I set the argument to be the new update message I put above, If I set it to twice, the test will fail cause when the first time it was called it was a different argument, for this line I was trying to make sure the updateTimesync did send out an updated message

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+package org.eclipse.mosaic.fed.infrastructure.ambassador;
+import java.io.IOException;
+
+import org.apache.commons.logging.Log;
+
+import com.google.gson.Gson;
+public class InfrastructureTimeInterface{
+public InfrastructureInstanceManager manager;
+//private int target_port=12123;
```
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java:25]: Will do, thank you for catching this

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * This function is used to send out encoded timestep update to all registered instances the manager has on the managed instances map
+ * 
+ * @param message This time message is used to store current seq and timestep from the ambassador side
+ * @throws IOException
+ */
+public void onTimestepUpdate(InfrastructureTimeMessage message) throws IOException 
+{
+    if(this.manager.get_managedInstances().size() == 0)
+    {
+        throw new IllegalAccessError("There are no registered instances!");
```
Comment (kjrush) [inline co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java:56]: I'm not sure that this is necessary, and may actually be a problem. In this case, if this is called every timestep then any timestep where we don't have any infrastructure instances registered would mean we either crash the infrastructure ambassador or the whole MOSAIC RTI. For scenarios where we're not using infrastructure simulation that would be a problem as well as any case where - due to the time it takes to launch the software - the infrastructure isn't able to connect on exactly the first timestep of simulation.

Regardless, if the role of this function is to send out messages to registered instances, then if there are 0 registered instances it seems straightforward to send 0 messages. Might be worthwhile to keep it as a log statement though, so we can catch if there are times where infrastructure is slow to register or somethin.

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    for(InfrastructureInstance current_instance: this.manager.get_managedInstances().values())
```
Comment (kjrush) [inline co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java:59]: I don't believe you need the `this` in `this.manager.get...`

File: co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package org.eclipse.mosaic.fed.infrastructure.ambassador;
```
Comment (kjrush) [inline co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java:1]: needs copyright statement.

**Review by kjrush — CHANGES_REQUESTED** [review]
Few minor things, before we merge I'd recommend running a format pass on things: https://code.visualstudio.com/docs/java/java-linting to cleanup indentation, variable naming, etc. if possible.

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureInstance.java [hunk 6]
```diff
@@ -71,4 +71,14 @@ public void sendMsgs(byte[] data) throws IOException {
         rxMsgsSocket.send(packet);
 
     }
+
+    public void sendTimesyncMsgs(byte[] data) throws IOException {
```
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureInstance.java:75]: Thank you Kyle, added documents comment to it

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureInstanceManager.java [hunk 7]
```diff
@@ -40,5 +40,10 @@
 public class InfrastructureInstanceManager {
     private Map<String, InfrastructureInstance>  managedInstances = new HashMap<>();
     private double currentSimulationTime;
+
+    public Map<String, InfrastructureInstance> get_managedInstances()
```
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureInstanceManager.java:44]: Thank you Kyle, changed all the function and variable naming to CamelCase

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureMessageAmbassador.java [hunk 8]
```diff
@@ -62,6 +63,9 @@ public class InfrastructureMessageAmbassador extends AbstractFederateAmbassador
     private InfrastructureTimeMessageReceiver InfrastructureTimeMessageReceiver;
     private Thread v2xTimeRxBackgroundThread;
     private InfrastructureInstanceManager InfrastructureInstanceManager = new InfrastructureInstanceManager();
+    private InfrastructureTimeInterface InfrastructureTimeInterface;
+
+    int Timesync_seq = 0;
```
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureMessageAmbassador.java:68]: Changed to private and CamelCase it

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureMessageAmbassador.java [hunk 9]
```diff
@@ -171,9 +175,16 @@ public synchronized void processTimeAdvanceGrant(long time) throws InternalFeder
...
             // TODO actions to do on queued Infrastructure  instance registration attempts
-
-            // TODO actions to do on queued Infrastructure  time sync messages
-
+           
+            Timesync_seq += 1;
+            InfrastructureTimeMessage timesync_message = new InfrastructureTimeMessage();
+            timesync_message.set_seq(Timesync_seq);
+            timesync_message.set_timestep(currentSimulationTime);
+            try {
```
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureMessageAmbassador.java:183]: Removed the try{}Catch{} inside , but I added another catch outside to catch the IOexception and logged it

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureMessageAmbassador.java [hunk 10]
```diff
@@ -62,6 +63,9 @@ public class InfrastructureMessageAmbassador extends AbstractFederateAmbassador
     private InfrastructureTimeMessageReceiver InfrastructureTimeMessageReceiver;
     private Thread v2xTimeRxBackgroundThread;
     private InfrastructureInstanceManager InfrastructureInstanceManager = new InfrastructureInstanceManager();
+    private InfrastructureTimeInterface InfrastructureTimeInterface;
```
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureMessageAmbassador.java:66]: Corrected the naming, Thank you for catching that Kyle.

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.IOException;
+
+import org.apache.commons.logging.Log;
+
+import com.google.gson.Gson;
+public class InfrastructureTimeInterface{
+public InfrastructureInstanceManager manager;
+//private int target_port=12123;
+//set to false on init release
+private Boolean await_infrastructure_advance_request = false;
```
[the remark this answers is not in the document]
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java:27]: Changed from Boolean to boolean, thank you for catching this Kyle

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java (hunk 3 above)
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java:56]: Removed the throw exception, added a log.info to state that there are no registered instances

File: co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java (hunk 4 above)
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/main/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterface.java:59]: Removed this, thank you Kyle

File: co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java (hunk 5 above)
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java:1]: Added copyright statement

File: co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java (hunk 1 above)
Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java:59]: added code to verify that the func is called twice

File: co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021 Old Dominion University. All rights reserved.
```
Comment (kjrush) [inline co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java:2]: Just noticed this, should be the Leidos copyright statement not the ODU one?

Comment (EricChen-Lei) [reply co-simulation/fed/mosaic-infrastructure/src/test/java/org/eclipse/mosaic/fed/infrastructure/ambassador/InfrastructureTimeInterfaceTest.java]: Changed to Leidos copyright

**Review by kjrush — APPROVED** [review]

**Pull request merged**